# Coding with AI 

This brief and very basic overview assumes that you know little-to-nothing about using **AI coding tools**
like Cursor, GitHub Copilot, Claude, or ChatGPT to build data analysis software.

You do **not** need to know anything about AI to start. Terms are defined as
will briefly explain what a model is, what a prompt is, what an agent does, what a context
file is, and so on. The goal is to make you a more effective user of these
tools so you get the productivity benefits *without* quietly introducing
errors into your data analysis pipelines. 

The tutorial also has an overview of when AI is likely to be most helpful, when it might not be so helpful, and the methodological responsibilities that come with using it. 

And for reference: this was written by js in late June, 2026...putting a date on it because it will probably be obselete by early July, 2026.

## 1. What is an AI coding assistant?

An AI coding assistant is a program that has consumed an enormous amount of text and
code and learned to predict plausible continuations and to build meta-awareness of a task so that it can effectively carry
out complex and multi-step operations. In simple use cases, you can describe what you
want (or start typing code), and the assistant will just suggest code that *statistically tends to follow*
in similar situations. In more sophisticated use cases, the AI will create its own "context" by trying to understand
what you want to do and then conditioning future outputs on that context. In all cases, the AI tool is 
optimizing for *plausible* code given the context of the task. This works amazingly well most of the time, but 
you sometimes get *plausible* code that runs, but that is not neccessarily *correct* in the context of your specific use case.
I will say up front that this is often the fault of the user, not the AI (the user's fault for not providing adequately constrained instructions). 
However, AIs can make mistakes on their own, especially when dealing with complex tasks. 

Typically, AI is used in three general forms:

* **Autocomplete (tab completion).** As you type, the tool suggests the rest of the
  line or block. You accept with Tab. Fast and easy to review.
* **Chat.** You ask a question in plain language ("how do I compute a PSTH?") and get
  an answer, often accompanied by code you can copy. Good for explanations and small snippets, and most of us
  probably use this all the time given the integration with Google search.
* **Agent.** You give a higher-level instruction ("add a function that loads the
  behavioral CSV and returns mean RT per condition") and the tool reads your files,
  writes code (often across many files), runs commands, and iterates until the task is complete.
  This is the most powerful use case and the one that needs the most supervision.

## 2. Other vocabulary terms

* **Model.** The underlying AI (e.g. Claude, GPT). Different models have
  different strengths, costs, and speeds. Bigger/newer is not automatically better
  for every task so do some homework before deciding. Some models excel at writing, others at
  coding, and some can do it all. Importantly, this decision also interacts heavily with *tokens and usage*
  (see below). 
* **Prompt.** What you ask. The quality of the output depends enormously on the
  quality of the prompt. Generally speaking, the more context and constraints you give, the better.
* **Context / context window.** The text the model can "see" at once such as your prompt,
  any open files, and recent conversations. It is finite. If your project is large,
  the model does not see all of it unless you point it at the relevant parts. Some environments,
  like Cursor, will summarize and compress the existing conversation/history when the context is full.
  This is good to avoid the model tracking old/obselete commands that might drive mistakes. 
* **Context files / rules.** A project file (e.g. a rules
  file or `guidelines.md`) describing your conventions such as "use numpy, label all axes,
  always set a random seed for reproducible outputs." These context files
  can be quite elaborate and can include information about writing style, being skeptical about
  claims, verifying facts using multiple sources, etc.
  *Thoughtfully writing out a context file is one of the most impactful steps you can take to obtain
  robust results (and to avoid some of the more annoying habits of AIs)*. Providing context also allows
  the assistant to always have access to the core concepts and this can reduce your context fingerprint.
* **Tokens and usage.** Models process text in chunks called tokens, and most
  services bill (or rate-limit) by tokens used. Long files and long conversations
  cost more. Practically, this means that you should keep your requests focused and use the lightest weight
  model that you can to achieve your goals. 
* **Hallucination.** When a model confidently produces something false such as a function
  that does not exist, a citation that was never written, a formula that is subtly
  wrong. Hallucinations happen and they are (so far) a fundamental property of
  how these systems work. Detecting them is *your* job. 

## 3. Brief set up guide

This tutorial is written with applications like **Cursor** in mind, which is an editor (a fork of VS Code) with AI
assistance built in. However, most of these principles apply to any tool (e.g., Claude Code, which is another favorite).

1. Install the editor and open your project **folder** (not just a single file), so
   the assistant can see all relevant data files and code.
2. Try the three main modes of operation using the default agent (Composer 2.5 as of June 2026): accept a tab-completion, ask a
   question in chat, and give the agent a small, well-scoped task.
4. Create a short **rules/context file** for the project stating your conventions
   (next section gives you some ideas for what to put in it).

Many tools offer free tiers or student/educator access, and I don't think anything in this tutorial depends on a paid plan.

## 4. How to work effectively with AI agents and some tips

Here are some habits that can help separate "AI made me faster" from "AI introduced a bug that my sci-enemy found
three months after I published."

* **Use version control.** Put your project in `git`. Commit working
  states often. Then, when the AI changes ten files, you can *see exactly what
  changed* (`git diff`) and undo it if needed. This single habit makes AI safe to
  experiment with. If you're editing a code file directly in cursor, it will also show
  you changes, but its good practice to keep backups as asking the wrong question can lead to massive
  code overhauls that can be challenging to decipher or revert. 
* **Keep raw data read-only.** Never let any script, written by AI-written or not, overwrite
  your raw data. Load it, copy it, transform the copy.
* **Work in small, reviewable steps.** Ask for one function, read it, test it, commit
  it. Giant "build the whole analysis" requests produce giant, unreviewable diffs (see point 1 above).
* **Point the model to the relevant  files** Use `@filename` to point the assistant to relevant files. This will
  allow you to control what the model sees (as opposed to having it read all the files in your working dir)
  and will reduce the percentage of context window currently in use. 
* **Give the model real context.** Tell it the *shape and units* of your data
  ("`X` is trials x timepoints, spikes in Hz"), the *design* ("within-subject, unequal
  trials"), and the *goal*. One tip: insert comments, espcially near data and loading data, with descriptions
  of the data / key ideas. That the model always has this information available, and you'll avoid many
  silent errors that arise when the model has to guess about these factors. Make sure to poke at the model every now and then
  (e.g. "When you wrote the main spike sorting function, what decisions were underspecified and required
  you to guess about the best approach?").
* **Teach the agent about your code** While you need to be mindful of context usage, have the agent read
  any relevant code and ask it to write a guide to tricky variables/the way things are saved,etc. As the agent
  to put that guide at the top of a new file, *before* having the agent help with analysis. This provides
  consistency, and helps to avoid common mistakes that occur when your naming conventions or style conflict with
  common usage. 
* **Keep track of the Context** In addition to context you provide, agents will seek additional context.
  This is both a good and bad: this ability makes them very powerful but it also means you may not be
  aware of it's understanding/state when it is answering. Using more to-the-metal models (e.g., Claude
  code vs. Claude co-work or Claude chat) helps because they tend to say more "outloud" to expose how
  they're thinking. This can help you build an accurate mental model of how they're going about the
  task.  
* **Context Drift** Related to last point. Context drift can take several forms and happens because AI's memory operates via token
  summary and compression, which neccessarily leads to distortions, especially of older details. For example,
  *conversation* drift happens when the model loses track of earlier decisions and develops a recency bias. In extreme
  cases this can lead to *requirement* (or *authority*) drift where a model can start treating its own outputs as
  more authoritative than your prompts (of course, you need to be very careful that you also don't start treating
  the model output as more authoritative than you are!). To combat, you can ask the model to periodically re-summarize its
  goals or to verify assumptions against context files.  
* **Ask it to explain, not just produce.** Related to point above: "Explain what this line does and why" is
  often more valuable than the code itself, and it can reveal wrong assumptions. Also remember that
  the agents are conditioning their current understanding on everything you've
  said and everything that the model has assumed so far. Starting "fresh" by clearing the context and
  *then* asking the agent to question everything is often more valuable than asking that in the current context.
  Another good trick is to ask the main agent to launch a subagent without the overall context and
  have the subagent read the experiment code & analysis code and look for mismatches - if you misled the main agent or
  it gets confused (context drift, etc), this 'monitoring subagent' approach will help flag it.
* **Make it reproducible.** Insist on tracking pinned package versions
  (`requirements.txt`), and code that runs top-to-bottom in a fresh kernel (i.e. that runs even if you kill
  the kernel, wipe everythign clean, and just press "go").
* **Read every line you keep.** If you would not be comfortable defending a line of
  code to your advisor or a collegue, or if would not be comfortable posting on a public repo, then do not
  commit it. "The AI wrote it" is not an excuse.

### Example of a simple context file:
A very simple but useful project rules file for a neuro/psych analysis might say (usually a markdown, or .md file): 

* use numpy and pandas to speed up numerical computations, avoid lists to store large arrays of data
* always average within subject before across subjects
* set a random seed and print it. Confirm that outputs are deterministic upon re-run
* label all plot axes with units
* never use exact `==` on floats
* add a comment above each functional block
* If a line of code is particulary Pythonic (e.g. list/dict comprehensions) or if it uses method chaining, then add a short comment to explain the sequence of operations
* **style** Don't number the comments or use all caps for emphasis.
* **style** Always use snake case. Camel case is aesthetically unbearable unless used for a class definition.
* **style** I am not formally trained in CSE, I am trained as a neuroscientist and a data scientist. I am very good at coding, but I prefer simple constructions and human-readable code that has easy-to-track flow. Avoid writing overly complicated code and we can optimize for performance after a working version is built.
  
The assistant will then apply these by default...and of course you can make this more elaborate to achieve better outcomes.
***The last style tip*** is one of my go-tos in a context file, because otherwise you can get mind-bendingly complicated code that i'm sure is beautiful to formally trained folks, but is a complete nightmare to someone like me. 


## 5. Code that runs but is wrong

Because they have digested an enormous amount of code, AI assistants are very good at producing code that *executes without error*. They are less reliable at producing code that is *scientifically correct* given your specific context/needs. The dangerous case is the intersection of
code that runs, returns a plausible number or a clean figure, and is **wrong**. *Often times this is due 
to user-error.* For example, having an underspecified context file, asking ambiguous questions, or not asking
for the right analysis will greatly increase the likelihood of functional, but wrong code (and the inverse is also true: writing clear and comprehensive context files and asking unambiguous questions will greatly mitigate this risk). 

Below are some toy examples where you'd get code that runs fine, but with underspecified context might not be correct.

### 5.1 This is not the t-test you are looking for

In [1]:
import numpy as np
from scipy import stats

rng = np.random.default_rng(0)

# two groups with unequal variances and sample sizes
group_a = rng.normal(500, 40, 30)
group_b = rng.normal(530, 120, 12)

# without more context, an assistant might assume the most commonnly encountered case and use an
# independent t-test with equal variance:
t_default, p_default = stats.ttest_ind(group_a, group_b)

# a better option for unequal variance (Welch's t-test):
t_welch, p_welch = stats.ttest_ind(group_a, group_b, equal_var=False)

print(f"default (Student, assumes equal variance): p = {p_default:.3f}")
print(f"Welch (does not assume equal variance):    p = {p_welch:.3f}")

default (Student, assumes equal variance): p = 0.024
Welch (does not assume equal variance):    p = 0.137


#### Solution
You need to spend some time with your data so you know what to ask for. Or, at minimum, you need to know the assumptions of the test you're asking for so you can prompt "If the variances are unequal by more than a factors of X%, use a Welch's t-test, otherwise use the standard Student's t-test". Of course this is a really simple example, and a quick (non-exhaustive) test revealed that a newer anthropic model immediately caught the unequal variance issue without explicit promting. However, it should be very easy to imagine what might happen if you knew nothing about the inner workings of a complex model (e.g., turning a DNN lose to decode your neural data). If you don't know the assumptions and the logic of what you're asking an assistant to do, then you shouldn't be asking the assistant to do it for you. 


### 5.2 Data leakage
* A slightly more complex example in the same general family of problem.
* Build a fake multi-feature data set that is pure noise, then run a logistic regression to decode task condition

In [4]:
from sklearn.model_selection import cross_val_score
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

rng = np.random.default_rng(0)

# 40 samples, 500 features, and labels that are noise (no real signal)
X = rng.normal(0, 1, (40, 500))
y = rng.integers(0, 2, 40)

# wrong approach: pick the features most correlated with y using all the data,
# then cross-validate on those hand-picked features
corrs = np.array([abs(np.corrcoef(X[:, j], y)[0, 1]) for j in range(X.shape[1])])
top = np.argsort(corrs)[-10:]
leaked = cross_val_score(LogisticRegression(max_iter=1000), X[:, top], y, cv=5).mean()

# right approach: do feature selection inside the cross-validation, via a pipeline
independent_pipe = make_pipeline(SelectKBest(f_classif, k=10),
                            LogisticRegression(max_iter=1000))
independent = cross_val_score(independent_pipe, X, y, cv=5).mean()

print(f"leaked 2-way decding accuracy (feature selection outside CV): {leaked:.2f}")
print(f"independent 2-way decoding accuracy (feature selection inside CV):  {independent:.2f}")

leaked 2-way decding accuracy (feature selection outside CV): 0.85
independent 2-way decoding accuracy (feature selection inside CV):  0.47


#### Solution
If you're not familar with the logic behind cross-validation/out-of-set prediction and the related topic of circular data analysis, then you might just ask an agent to "select the most informative features to maximize decoding accuracy". Again, a smart agent might know to use independent training/validation splits, but you wouldn't be sure it was right (esp if you were super excited with your 85% decoding accuracy!).   

## 6. Prompts and review habits that reduce risk

You can't prompt your way to 100% correctness, but good prompts and review
habits can dramatically lower the error rate:

* **Clearly state the parameters of an ask in the prompt.** "Within-subject design, unequal trials per
  subject; average within subject first." "Variances are unequal." "Correct for
  multiple comparisons across 64 channels." "Make sure that data used to train a model or to select hyperparameters
  is completely independent from the validation data." If you do not say it, the model has to guess. It might
  guess right, it might guess wrong. But if you don't know what you're doing, then you can't verify. 
* **Ask about any the assumptions.** "List the statistical assumptions this code makes and
  highlight steps that required you to guess about assumption or where the analysis is most likely to be wrong."
  This will prompt the model to generate a checklist that you can verify (again, *only if you know what you're doing!*).
* **Request a test or a sanity check.** "Add an assertion that the array is
  trials x time" or "show me the result on a tiny example where I know the answer."
* **Diff and review.** Read `git diff` for every AI change. If a change touches a file
  you did not expect, stop and look (I have recently seen this happen, even with frontier models).
* **Verify against a known result.** Reproduce a number from a paper, or compute a
  toy case by hand, before trusting the pipeline on real data.
* **Prefer boring, explicit code.** If the AI writes something clever you do not
  fully understand, ask it to rewrite it simply. Ultimately, you are 100% responsible
  for any errors (see the example context file above for a sample prompt related to simple code).

## 7. Minimal checklist

Before you trust AI-written analysis code, confirm:

* The project is in `git` and you have reviewed the `diff`.
* Raw data is untouched. All analyses and transformations act on copies.
* Array axes and data shapes are what you think (`assert` them).
* If appropriate, random seeds are set and reported.
* All package versions are recorded for reproducibility.
* Statistical choices (test, error bars, corrections, exclusions) are *your*
  deliberate decisions, documented in comments and the methods section.
* You validated the pipeline against a known/hand-computed result. I **highly** recommend
  generating simulated data where you know ground truth. Then run the simulated data through the pipeline
  to make sure you get back the ground truth.
* You can explain every line in the code. I'll do an AI thing here and use all caps: NEVER publish
  anything based on code that you don't 100% understand. That will for sure come back to bite you. Indeed,
  it has never been easier to find erorrs in other people's code!

## 8. Pros, cons, and methodological/ethical considerations

This section reflects my opinion, your opinion may vary. 
AI coding tools are neither a magic productivity button nor a threat to be banned.
They can be a powerful tool, and using them
responsibly in science means understanding how they work and most importantly, taking as much personal responsibility
for their output as you would for code that your wrote entirely on your own.

### Where AI can help the most

* **Lowering the barrier to entry.** A researcher with a solid
  question but limited programming background can get started quickly, prototype an
  analysis, and learn by reading explanations provided by the agent. This is cool, and
  I've had many fun conversations with agents that have revealed novel approaches that
  I wouldn't have thought of otherwise.
* **Doing highly stereotyped tasks.** Loading standard file types, reshaping arrays, generating plots,
  writing argument parsers, etc. These are tedious, but neccessary, tasks that are slow to type.
  AI is usually excellent here, as they have consumed tons and tons of examples of these common tasks.  
* **Explanation and learning.** An AI can be a patient tutor that explains unfamiliar concepts,
  and unlike your advisor or your friends, it is available at 3 a.m. when you wake up with a cool idea. 
* **Refactoring and documentation.** Turning a working-but-messy script into clear,
  commented, tested code base that runs effeciently. AIs are also great at suggesting edge cases
  that you might have missed.
* **Breadth.** Finding libraries and algorithms that you didn't know existed.

### Where AI can hurt

* **Confident wrongness.** As Section 5 showed using super simple examples, the characteristic failure is code
  that runs and looks right but is incorrect given a specific context. This is more dangerous
  than code that crashes, because nothing will flag these errors.
* **Plausible fabrication.** Models can invent function names, arguments, citations, and
  statistical justifications that do not exist. Surely this will become less and less common,
  but it happens (and see note about context drift in section `4`). 
* **Reproducing biases.** Models reproduce the most common patterns in
  their training data, including common *mistakes* (e.g. uncorrected
  multiple comparisons, leaky cv pipelines). As your analyses get more complex, these biases
  can be really hard to identify. Ultimately they learn from our artifacts (mostly), so to the extent humans are
  consistently biased, these models will also be biased.  
* **Skill atrophy and shallow understanding.** If you never struggle through writing
  an analysis, you will not develop the judgment needed to recognize when the analysis
  is wrong. I know how tempting it is to just turn these things lose on a problem, but if you
  don't struggle through learning stats, coding, etc, you're setting yourself up for disaster. 
* **Homogenization.** If everyone accepts the same suggested approach, and that approach is non-optimal or
  wrong, then it potentially sets up a cycle that exacerbates the 'biases' problem mentioned above. 

### Methodological/Ethical considerations

These are the issues to think about *before* AI-generated code ends up in a a paper:

* **Reproducibility and provenance.** Reproducibility is a property of the *code and
  data*, not of who or what wrote them. However, this requires discipline and careful tracking
  (e.g., all packages, assumptions, etc). Keep a detailed record of what you ran. Some
  groups note in their methods that AI assistance was used. These norms are still forming,
  and you should follow your field's and institution's emerging guidance.
* **Validation.** This sounds harsh, but treat AI-written analysis code the way you would
  treat code given to you by someone who you didn't think was very good at coding :). Calibrate it against
  known results before you trust it. The burden of correctness is on you, not on the AI. 
* **Transparency in reporting.** Your methods section must describe what the analysis
  actually does, including the tests, corrections, exclusion criteria, and parameters. There must be enough
  detail to reproduce the findings, regardless of how the code was written. "We used an AI
  generated script" is not a method.
* **Data privacy, ethics, and the IRB.** Sending data or code to a cloud AI service
  can violate participant consent, data-use agreements, IRB protocols, HIPAA/FERPA,
  or GDPR, **especially with human-subjects data containing identifiers**. Never paste
  participant data (or anything personally identifying) into an external AI tool
  unless you are certain it is permitted. When in doubt, ask your PI or IRB first, but generally
  only use local/offline tools for sensitive data.
* **Equity of access.** The best tools cost money, sometimes lots of money. Be mindful that AI assistance can
  widen gaps between well-resourced and under-resourced labs and students.
* **Environmental cost.** Large models consume substantial energy. Use them deliberately rather than reflexively.